### PR TITLE
doc/103: Documented conversions

### LINKED ISSUE #103: Document conversions
Document how the different mappings convert different input types.

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #103

### CODE REVIEW

File: doc/user_guide/edml_user_guide.md [hunk 1]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+* Nested object: Not convertable
+* Nested list: Not convertable
+* Decimal value: String representation of decimal (e.g: `"1.23"`)
+* Double value: String representation of decimal (e.g: `"1.23"`)
+* Boolean value: `"true"` or `"false"`
+* Binary data: Base64 encoded
+* Date: Date as string (e.g: `"2021-09-27"`)
+* Timestamp: Timestamp as UTC timestamp (e.g: `"2021-09-21T08:18:38Z"`)
+
+For details, you can also check the implementation of the converter: [PropertyToVarcharColumnValueExtractor.java](../../src/main/java/com/exasol/adapter/document/mapping/PropertyToVarcharColumnValueExtractor.java)
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:263]:
```suggestion
```

Comment (morazow) [reply doc/user_guide/edml_user_guide.md:263]: I think we can remove this. If it is required, we could add it to developer guide.

File: doc/user_guide/edml_user_guide.md [hunk 2]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+
+### ToDecimalMapping Conversions
+
+* Nested object: Not convertable
+* Nested list: Not convertable
+* String: The adapter tries to parse the string as a number. E.g: `"1.23"` -> `1.23`. If not possible (e.g: `"abc"`) the adapter handles the value as not convertable.
+* Decimal value: No conversion needed
+* Double value: converted to decimal
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:271]:
```suggestion
* Double value: Converted to decimal
```

File: doc/user_guide/edml_user_guide.md [hunk 3]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+* `CONVERT_OR_ABORT` try to convert and abort the query if not convertable
+* `CONVERT_OR_NULL` try to convert and return `NULL` if not convertable
+
+All mappings pass through null values. That means, if the source value is a null-value, the adapter converts it to `NULL`. The only exception is the `toJsonMapping` which converts null values to JSON null values.
+
+### ToVarcharMapping Conversions
+
+* String: No conversion needed
+* Nested object: Not convertable
+* Nested list: Not convertable
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:255]:
```suggestion
* Nested object: Not convertable
* Nested list: Not convertable
* String: No conversion needed
```

File: doc/user_guide/edml_user_guide.md [hunk 4]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+### ToDecimalMapping Conversions
+
+* Nested object: Not convertable
+* Nested list: Not convertable
+* String: The adapter tries to parse the string as a number. E.g: `"1.23"` -> `1.23`. If not possible (e.g: `"abc"`) the adapter handles the value as not convertable.
+* Decimal value: No conversion needed
+* Double value: converted to decimal
+* Boolean value: `true` -> 1, `false` -> 0
+* Binary data: Not convertable
+* Date: Date as UTC timestamp in milliseconds
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:274]: Maybe good idea to add, if decimal or double value floor or ceiling used for date and timestamp conversions?

Comment (morazow) [inline doc/user_guide/edml_user_guide.md:274]:
```suggestion
* Date: Date as UTC milliseconds time value
```

File: doc/user_guide/edml_user_guide.md [hunk 5]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+### ToDoubleMapping Conversions
+
+* Nested object: Not convertable
+* Nested list: Not convertable
+* String: The adapter tries to parse the string as a number. E.g: `"1.23"` -> `1.23`. If not possible (e.g: `"abc"`) the adapter handles the value as not convertable.
+* Decimal value: Converted to floating-point.
+* Double value: No conversion needed
+* Boolean value: `true` -> 1, `false` -> 0
+* Binary data: Not convertable
+* Date: Date as UTC timestamp in milliseconds
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:286]:
```suggestion
* Date: * Date: Date as UTC milliseconds time value
```

File: doc/user_guide/edml_user_guide.md [hunk 6]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+### ToJsonMapping Conversions
+
+The `toJsonMapping` always converts the input value to a JSON string. For that reason there is no property like `nonStringBehaviour`.
+
+* Nested object: Converted to JSON string
+* Nested list: Converted to JSON string
+* String: Converted to JSON string
+* Decimal value: Converted to JSON number
+* Double value: Converted to JSON number
+* Boolean value: converted to JSON boolean
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:322]:
```suggestion
* Boolean value: Converted to JSON boolean
```

File: src/main/java/com/exasol/adapter/document/mapping/AbstractPropertyToNumberColumnValueExtractor.java [hunk 7]
```diff
@@ -133,12 +133,16 @@ public void visit(final DocumentBinaryValue binaryValue) {
...
+                    .convertDecimal(BigDecimal.valueOf(dateValue.getValue().getTime()));
+            this.result = handleNotNumericButConvertAble(converted, "<date>");
         }
 
         @Override
         public void visit(final DocumentTimestampValue timestampValue) {
-            this.result = handleNotNumeric("<timestamp>");
+            final Object converted = this.numberConverter
+                    .convertDecimal(BigDecimal.valueOf(timestampValue.getValue().getTime()));
+            this.result = handleNotNumericButConvertAble(converted, "<timestamp>");
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/AbstractPropertyToNumberColumnValueExtractor.java:145]:
```suggestion
            this.result = handleNotNumericButConvertable(converted, "<timestamp>");
```

File: src/main/java/com/exasol/adapter/document/mapping/AbstractPropertyToNumberColumnValueExtractor.java [hunk 8]
```diff
@@ -133,12 +133,16 @@ public void visit(final DocumentBinaryValue binaryValue) {
...
         }
 
         private Object handleNotNumericButConvertAble(final Object converted, final String value) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/mapping/AbstractPropertyToNumberColumnValueExtractor.java:148]:
```suggestion
        private Object handleNotNumericButConvertable(final Object converted, final String value) {
```

File: doc/user_guide/edml_user_guide.md [hunk 9]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+* Binary data: Converted to JSON string with base 64 encoded data
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:323]:
```suggestion
* Binary data: Converted to JSON string with Base64 encoded data
```

File: doc/user_guide/edml_user_guide.md [hunk 10]
```diff
@@ -229,10 +229,101 @@ Some dialects support reading one table from multiple sources. For example the [
...
+
+### ToVarcharMapping Conversions
+
+* String: No conversion needed
+* Nested object: Not convertable
+* Nested list: Not convertable
+* Decimal value: String representation of decimal (e.g: `"1.23"`)
+* Double value: String representation of decimal (e.g: `"1.23"`)
+* Boolean value: `"true"` or `"false"`
+* Binary data: Base64 encoded
```
Comment (morazow) [inline doc/user_guide/edml_user_guide.md:259]:
```suggestion
* Binary data: Converted to Base64 encoded data string
```

**Review by morazow — CHANGES_REQUESTED** [review]

File: doc/user_guide/edml_user_guide.md (hunk 4 above)
Comment (jakobbraun) [reply doc/user_guide/edml_user_guide.md]: like that?

**Review by morazow — APPROVED** [review]

**Pull request merged**